packages/engine/lib/execution/src/package/mod.rs

```diff
@@ -221,11 +221,10 @@ pub trait PackageCreator: Send + Sync {
     /// A message sent to all workers before running any packages.
     ///
     /// This allows package creators to pass any kind of configuration from their Rust runtime to
-    /// their Language Runner counterpart all.
+    /// their Language Runner counterpart for the experiment.
     ///
     /// Compared to [`Package::start_message()`], the data returned with this method will be
-    /// available for all simulations. Also, this should not be implemented for packages but
-    /// their respective package creator.
+    /// available for all simulations and is sent **once per experiment**.
```

Comment:
I still think the wording (and naming of the methods) isn't clear. Lets try and rethink this from a high-level.

These methods are implemented by Packages (and their Creators) to define how they register themselves with the Language Runners, and to provide any initialisation data that they need to. The one on the `PackageCreator` is sent at the start of the experiment. (This is why it's on the `PackageCreator` because `Package`s are scoped to a Simulation but the Creators, i.e. factories, are unique)

The one on the `Package` is sent at the start of a simulation by a Package when it's created. This allows the Package to send initialisation data to the language runners.

Do you know where the messages actually go? As in what code they reach within the language runners, I'm assuming the package impls?

Reply:
The `init_message` method is only used for behavior execution and is used to create `Vec<BehaviorDescription>`:
```rust
pub struct BehaviorDescription {
    pub id: BehaviorId,
    pub name: String,
    pub short_names: Vec<String>,
    pub source: String,
    pub required_field_keys: Vec<String>,
    pub language: Language,
    pub dyn_access: bool,
}
```

It's encapsulated in `ExperimentInitRunnerMsgBase`, which contains general information about the experiment, which is passed to `spawn_worker`.

The `start_message` isn't used at all currently. The messages are passed with `NewSimulationRun` to all runners when a new simulation starts.
```rust
pub struct NewSimulationRun {
    pub span: Span,
    pub short_id: SimulationShortId,
    pub worker_allocation: Arc<WorkerAllocation>,
    pub packages: PackageMsgs,
    pub datastore: DatastoreSimulationPayload,
    pub globals: Arc<Globals>,
}
```

Any naming ideas? I think this high level overview should simplify the naming a little bit. The former is used like a `worker_init_message` and the latter is used as a `simulation_start_message`.

Reply:
I think those suggestions are good. (Isn't that basically what they were before? 😁 )